### DISCUSSION TITLE
kustomize-controller: "validation failed: error: no objects passed to apply"

### DISCUSSION THREAD

**Discussion by cubbasa** [fluxcd/flux2#382] [Q&A] [answered]
I'm using single plain yaml manifest file (with multiple objecst in one manifest file) stored in git or bucket.
There is single kustomize object with source object (from git or bucket) to apply changes to cluster.
When I've made changes inside manifest files by adding or removing K8s objects and at least one object is left in manifest everything working as expected.
According to documentation (https://toolkit.fluxcd.io/components/kustomize/kustomization/#garbage-collection)
_Garbage collection means that the Kubernetes objects that were previously applied on the cluster but are missing from the current source revision, are removed from cluster automatically. Garbage collection is also performed when a Kustomization object is deleted, triggering a removal of all Kubernetes objects previously applied on the cluster._
When manifest is removed (deleted) from source (git or bucket) or it become empty error is thrown by Kustomize controller when it try to reconcile state from git or bucket to cluster: _"validation failed: error: no objects passed to apply"_  and all objects are left in cluster without deletion. The only way to recover from this state is to remove Kustomize object that apply changes from git or bucket to cluster.

My question is that the way how it should work that only way to delete all objects in cluster described by manifest is to remove Kustomize object? I've got situations that I would like to have Kustomize object(s) only managed by cluster or namespace admins (like avoid exposing of credentials to git or bucket) and don't allow devs to view or change this object but I need to leave option to remove app(s) by them if they needed it (like by removing manifest from repo or bucket which they have access to) and recreate it but with current way how it works is not possible. 
My following question is if it possible to allow it to work as I wrote earlier that by removing manifest from source repository or bucket to trigger deletion of all objects that were described by removed manifest and applied to cluster?

**Comment by kingdonb** [comment, marked as answer]
Thanks for your question. Flux treats an empty kustomization as invalid, because I think we assume this is an error. Your use case makes sense to me but it is not currently supported in the way that you imagined.

The devs will have to leave at least one resource in the deflated Kustomization/GitRepo, perhaps such as a configmap, as otherwise the validation will fail seeing no resources to apply, and their resources will not be deleted as they expected.
